File: src/migration_rules/format_string_rule/after_script.py
#!/usr/bin/env python3
"""
Format String Migration - AFTER Script (Spark 3.3+)
This script demonstrates the correct 1$-based indexing in format_string and printf functions.

Usage:
    python after_script.py

This script can be run independently to test the corrected behavior.
"""

from pyspark.sql import SparkSession
from pyspark.sql.functions import expr

def main():
    # Create SparkSession
    spark = SparkSession.builder \
        .appName('FormatStringAfter') \
        .master('local[*]') \
        .getOrCreate()
    
    print("=== Format String Migration - AFTER (Correct 1$-based indexing) ===")
    
    try:
        # Create sample DataFrame
        data = [
            ('John', 5, 'Premium'),
            ('Alice', 100, 'Gold'),
            ('Bob', 25, 'Silver'),
            ('Carol', 75, 'Premium')
        ]
        df = spark.createDataFrame(data, ['name', 'message_count', 'tier'])
        
        print("\nOriginal DataFrame:")
        df.show()
        
        # Using format_string with 1$-based indexing (correct for Spark 3.3+)
        print("\n1. Using format_string with 1$-based indexing:")
        result1 = df.select(
            'name',
            'message_count',
            expr("format_string('Hello %1$s, you have %2$d messages', name, message_count)").alias('greeting')
        )
        result1.show(truncate=False)
        
        # Using printf with 1$-based indexing (correct for Spark 3.3+)
        print("\n2. Using printf with 1$-based indexing:")
        result2 = df.select(
            'name',
            'tier',
            expr("printf('User %1$s has %2$s tier status', name, tier)").alias('user_info')
        )
        result2.show(truncate=False)
        
        # Complex format string with proper argument indexing
        print("\n3. Complex format string with proper 1$-based indexing:")
        result3 = df.select(
            'name',
            'message_count',
            'tier',
            expr("format_string('Welcome %1$s! You are a %3$s member with %2$d unread messages', name, message_count, tier)").alias('welcome_msg')
        )
        result3.show(truncate=False)
        
        # Demonstrating argument reuse with proper indexing
        print("\n4. Argument reuse with proper indexing:")
        result4 = df.select(
            'name',
            expr("format_string('Hello %1$s, welcome back %1$s!', name)").alias('reuse_example')
        )
        result4.show(truncate=False)
        
        # Mixed format types with proper indexing
        print("\n5. Mixed format types with proper indexing:")
        result5 = df.select(
            'name',
            'message_count',
            'tier',
            expr("printf('User: %1$s | Messages: %2$d | Tier: %3$s | Score: %2$d%%', name, message_count, tier)").alias('detailed_info')
        )
        result5.show(truncate=False)
        
        # Nested format string operations with correct indexing
        print("\n6. Nested format string operations with 1$-based indexing:")
        result6 = df.select(
            'name',
            'message_count',
            expr("format_string('Status: %1$s', format_string('User %1$s has %2$d items', name, message_count))").alias('nested_format')
        )
        result6.show(truncate=False)
        
        print("\n✅ SUCCESS: All format string operations completed successfully!")
        print("   This code uses correct 1$-based indexing compatible with Spark 3.3+")
        print("   All argument references start from 1$ instead of 0$")
        
    except Exception as e:
        print(f"❌ Error executing format string operations: {e}")
        print("   If you see this error, there might be an issue with the Spark setup or syntax")
    
    finally:
        spark.stop()
        print("\nSpark session stopped.")

if __name__ == "__main__":
    main()

File: src/migration_rules/format_string_rule_scala/after_script.scala
// Format String Argument Index Migration - AFTER Script (Spark 3.3+)
// This script demonstrates the updated behavior where format_string() and printf() require 1$-based indexing.
//
// Language: Scala
// Usage: spark-shell -i after_script.scala
//
// This script can be run independently to test the updated behavior.

import org.apache.spark.sql.SparkSession
import org.apache.spark.sql.functions._
import org.apache.spark.sql.types._

object FormatStringMigrationAfter {
  def main(args: Array[String]): Unit = {
    // Create SparkSession
    val spark = SparkSession.builder()
      .appName("FormatStringMigrationAfter")
      .master("local[*]")
      .config("spark.sql.warehouse.dir", "spark-warehouse")
      .getOrCreate()

    import spark.implicits._

    println("=== Format String Argument Index Migration - AFTER (Spark 3.3+ behavior) ===")

    try {
      // Create sample data
      println("\n📊 Creating sample data...")
      val data = Seq(
        ("Alice", 25, 50000.0),
        ("Bob", 30, 60000.0),
        ("Charlie", 35, 70000.0)
      ).toDF("name", "age", "salary")

      data.show()

      // Demonstrate that 0$ indexing no longer works
      println("\n🔍 ATTEMPTING: Using 0$ indexing (should fail in Spark 3.3+)...")
      println("   SQL: SELECT format_string('Employee: %0$s, Age: %1$d', name, age)")

      try {
        val result1 = data.select(
          col("name"),
          col("age"),
          expr("format_string('Employee: %0$s, Age: %1$d', name, age)").alias("formatted_0_index")
        )

        result1.show(truncate = false)
        println("❌ UNEXPECTED: 0$ indexing worked (this shouldn't happen in Spark 3.3+)")

      } catch {
        case e: Exception =>
          println(s"✅ EXPECTED FAILURE: ${e.getMessage}")
          println("   🎉 Spark 3.3+ correctly rejects 0$ indexing!")
      }

      // Demonstrate the CORRECT approach with 1$-based indexing
      println("\n✅ CORRECT APPROACH: Using 1$-based indexing...")
      println("   SQL: SELECT format_string('Employee: %1$s, Age: %2$d', name, age)")

      try {
        val result2 = data.select(
          col("name"),
          col("age"),
          expr("format_string('Employee: %1$s, Age: %2$d', name, age)").alias("formatted_1_index")
        )

        println("✅ SUCCESS: 1$-based indexing works correctly")
        result2.show(truncate = false)

      } catch {
        case e: Exception =>
          println(s"❌ Unexpected error with correct indexing: ${e.getMessage}")
      }

      // Demonstrate correct printf usage
      println("\n✅ CORRECT: Using 1$-based indexing in printf...")
      println("   SQL: SELECT printf('Salary: %1$.2f for %2$s', salary, name)")

      try {
        val result3 = data.select(
          col("name"),
          col("salary"),
          expr("printf('Salary: %1$.2f for %2$s', salary, name)").alias("printf_1_index")
        )

        println("✅ SUCCESS: printf with 1$-based indexing")
        result3.show(truncate = false)

      } catch {
        case e: Exception =>
          println(s"❌ Error: ${e.getMessage}")
      }

      // Show complex example with proper indexing
      println("\n✅ BEST PRACTICE: Complex format string with 1$-based indexing...")
      println("   SQL: SELECT format_string('%1$s is %2$d years old and earns %3$.2f. %1$s works hard!', name, age, salary)")

      try {
        val result4 = data.select(
          col("*"),
          expr("format_string('%1$s is %2$d years old and earns %3$.2f. %1$s works hard!', name, age, salary)")
            .alias("complex_1_index")
        )

        println("✅ SUCCESS: Complex 1$-based indexing")
        result4.show(truncate = false)

      } catch {
        case e: Exception =>
          println(s"❌ Error: ${e.getMessage}")
      }

      // Demonstrate DataFrame API with correct indexing
      println("\n📊 DataFrame API with correct 1$-based indexing...")
      try {
        val dfResult = data.withColumn(
          "description",
          expr("format_string('Name: %1$s, Details: Age %2$d, Salary $%3$.2f', name, age, salary)")
        )

        println("✅ DataFrame API with 1$-based indexing:")
        dfResult.select("name", "description").show(truncate = false)

      } catch {
        case e: Exception =>
          println(s"❌ DataFrame API error: ${e.getMessage}")
      }

      // Show alternative approaches without explicit indexing
      println("\n💡 ALTERNATIVE: Using format strings without explicit indexing...")
      try {
        val result5 = data.select(
          col("*"),
          expr("format_string('Employee: %s, Age: %d, Salary: %.2f', name, age, salary)")
            .alias("simple_format")
        )

        println("✅ SUCCESS: Simple format without explicit indexing")
        result5.show(truncate = false)

      } catch {
        case e: Exception =>
          println(s"❌ Error: ${e.getMessage}")
      }

      // Demonstrate concat_ws as an alternative
      println("\n💡 ALTERNATIVE: Using concat_ws for simple string concatenation...")
      try {
        val result6 = data.select(
          col("*"),
          concat_ws(" - ", 
            concat(lit("Name: "), col("name")),
            concat(lit("Age: "), col("age")),
            concat(lit("Salary: $"), format_number(col("salary"), 2))
          ).alias("concat_alternative")
        )

        println("✅ SUCCESS: concat_ws alternative")
        result6.show(truncate = false)

      } catch {
        case e: Exception =>
          println(s"❌ Error: ${e.getMessage}")
      }

      println("\n🎯 MIGRATION RECOMMENDATIONS:")
      println("   ✅ Replace all %0$ references with %1$")
      println("   ✅ Increment all other index numbers by 1")
      println("   ✅ Use simple format strings without indexing when possible")
      println("   ✅ Consider concat_ws() for simple concatenations")
      println("   ✅ Test all format_string() and printf() calls after upgrade")

    } catch {
      case e: Exception =>
        println(s"❌ Error during demonstration: ${e.getMessage}")
        println("   Check your Spark version and configuration")
    } finally {
      spark.stop()
      println("\nSpark session stopped.")

      println("\n🎉 SUMMARY - Spark 3.3+ Benefits:")
      println("   • Consistent with standard printf implementations")
      println("   • Eliminates confusion about argument indexing")
      println("   • Prevents off-by-one errors")
      println("   • Improves code readability and maintainability")
    }
  }
}

// If running in spark-shell, call the main method
FormatStringMigrationAfter.main(Array())